test/src/test/java/org/apache/accumulo/test/functional/TableChangeStateIT.java

```diff
@@ -0,0 +1,549 @@
+
+  @Override
+  protected int defaultTimeoutSeconds() {
+    return 4 * 60;
+  }
+
+  /**
+   * Validate that {@code TableOperations} online operation does not block when table is already online and fate transaction lock is held by other operations.
+   * The test creates, populates a table and then runs a compaction with a slow iterator so that operation takes long enough to simulate the condition. After
+   * the online operation while compaction is running completes, the test is complete and the compaction is canceled so that other tests can run.
+   *
+   * @throws Exception
+   *           any exception is a test failure.
+   */
+  @Test
+  public void changeTableStateTest() throws Exception {
+
+    Connector connector = getConnector();
+    String tableName = getUniqueNames(1)[0];
+
+    createData(tableName);
+
+    assertEquals("verify table online after created", TableState.ONLINE, getTableState(connector, tableName));
+
+    OnlineStatus status = onLine(tableName);
+
+    log.trace("Online 1 in {} ms", TimeUnit.MILLISECONDS.convert(status.runningTime(), TimeUnit.NANOSECONDS));
+
+    assertFalse("verify no exception thrown changing state", status.hasError());
+    assertEquals("verify table is still online", TableState.ONLINE, getTableState(connector, tableName));
+
+    // verify that offline then online functions as expected.
+
+    connector.tableOperations().offline(tableName, true);
+    assertEquals("verify table is offline", TableState.OFFLINE, getTableState(connector, tableName));
+
+    OnlineStatus status2 = onLine(tableName);
+
+    log.trace("Online 2 in {} ms", TimeUnit.MILLISECONDS.convert(status2.runningTime(), TimeUnit.NANOSECONDS));
+
+    if (status2.hasError()) {
+      log.debug("Online failed with exception", status2.getException());
+    }
+
+    assertFalse("verify no exception thrown changing state", status2.hasError());
+    assertEquals("verify table is back online", TableState.ONLINE, getTableState(connector, tableName));
+
+    // launch a full table compaction with the slow iterator to ensure table lock is acquired and held by the compaction
+    Thread compactThread = new Thread(new SlowCompactionRunner(tableName));
+    compactThread.start();
+
+    assertTrue("verify that compaction running and fate transaction exists", blockUntilCompactionRunning());
+
+    // try to set online while fate transaction is in progress - before ACCUMULO-4574 this would block
+    OnlineStatus status3 = onLine(tableName);
+    log.trace("Online while compacting in {} ms", TimeUnit.MILLISECONDS.convert(status3.runningTime(), TimeUnit.NANOSECONDS));
+
+    if (status3.hasError()) {
+      log.debug("Online failed with exception", status3.getException());
+    }
+
+    assertFalse("verify no exception thrown changing state", status3.hasError());
+    assertTrue("online should take less time than expected compaction time", status3.runningTime() < TimeUnit.NANOSECONDS.convert(2, TimeUnit.SECONDS));
```

Comment:
Increased from 2 seconds to be < compaction time which is 100 seconds and can be modified by changing the number of data rows or scanner sleep time, (currently 1000 rows and 100 ms sleep.)  On my VM this test is taking ~5 ms.